src/main/java/com/google/api/generator/engine/ast/ArithmeticOperationExpr.java

```diff
@@ -0,0 +1,93 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.api.generator.engine.ast;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+
+@AutoValue
+public abstract class ArithmeticOperationExpr implements OperationExpr {
+
+  public abstract Expr lhsExpr();
+
+  public abstract Expr rhsExpr();
+
+  public abstract OperatorKind operatorKind();
+
+  public abstract TypeNode type();
+
+  @Override
+  public void accept(AstNodeVisitor visitor) {
+    visitor.visit(this);
+  }
+
+  public static ArithmeticOperationExpr concatWithExprs(Expr lhsExpr, Expr rhsExpr) {
+    return builder()
+        .setLhsExpr(lhsExpr)
+        .setRhsExpr(rhsExpr)
+        .setOperatorKind(OperatorKind.ARITHMETIC_ADDITION)
+        .setType(TypeNode.STRING)
+        .build();
+  }
+
+  private static Builder builder() {
+    return new AutoValue_ArithmeticOperationExpr.Builder();
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+
+    // Private setter
+    abstract Builder setLhsExpr(Expr expr);
+
+    // Private setter
+    abstract Builder setRhsExpr(Expr expr);
+
+    // Private setter
+    abstract Builder setOperatorKind(OperatorKind operator);
+
+    // Private setter
+    abstract Builder setType(TypeNode type);
+
+    abstract ArithmeticOperationExpr autoBuild();
+
+    private ArithmeticOperationExpr build() {
+      ArithmeticOperationExpr arithmeticOperationExpr = autoBuild();
+      TypeNode lhsExprType = arithmeticOperationExpr.lhsExpr().type();
+      TypeNode rhsExprType = arithmeticOperationExpr.rhsExpr().type();
+      OperatorKind operator = arithmeticOperationExpr.operatorKind();
+      final String errorMsg =
+          String.format(
+              "Arithmetic operator %s can not be applied to %s, %s.",
+              operator, lhsExprType.toString(), rhsExprType.toString());
+
+      // None of expression should be void type
+      Preconditions.checkState(
+          !lhsExprType.equals(TypeNode.VOID) && !rhsExprType.equals(TypeNode.VOID), errorMsg);
```

Comment:
Instead of checking for `void`, what about checking that they are valid types instead? Note that `int + String` is not valid but `String + int` (or most things on the RHS, I think, please double-check this) is fine.

If we want to (optionally) support numeric types, we can check that here too, and then call `setType` with the appropriately widened type. This is a P2 though, so feel free to leave it for a separate PR.

Reply:
For concat case, one of expr is string-type is good regardless the position and none of expr is void type. Saying like `5 + "abc"` and `"abc" + 5` are both valid.

Regarding to `void`, for all arithmetic operation (+,-,*,/,%) the expression type on lhs and rhs have to be non-void type. So I think it is worth to take it out for all operators. 

In my original design thinking, we could separate addition operator usage into concat and plus. Because concat always return String, plus return type is based on expression type and related to data-type precedence, like
```
int a = 5; double b = 5.5;
double c = a + b;
```
when we need to support numeric plus usage, we might have another convenience method like `plusWithExprs`, different type-checking and `setType` for plus usage.

Therefore, void type-checking for all operators, and for each operator/usage has its own type-checking. LMK if you have other opinion. Thanks :)